<a href="https://colab.research.google.com/github/Vitelmojr01/REPOSITORIO-04/blob/modulo-regressaoII-modulo-13-tarefa-01/Mod13_Tarefa_I_fazendo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('previsao_de_renda.csv')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   id_cliente             15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  renda                  15000 non-null  float64
dtypes:

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.

2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.

3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor.
    

Resolvendo Exercicio 01 - item 01

In [ ]:
import pandas as pd, numpy as np, statsmodels.formula.api as smf
df = pd.read_csv('previsao_de_renda.csv')
print("Base carregada:", df.shape)

df = df.drop(columns=['Unnamed: 0','data_ref','id_cliente'])
print("Colunas após limpeza:", df.columns.tolist())

modelo1 = smf.ols("""np.log(renda) ~ posse_de_veiculo + posse_de_imovel + qtd_filhos + idade +
tempo_emprego + qt_pessoas_residencia +
C(sexo, Treatment(reference='F')) +
C(tipo_renda, Treatment(reference='Assalariado')) +
C(educacao, Treatment(reference='Secundário')) +
C(estado_civil, Treatment(reference='Casado')) +
C(tipo_residencia, Treatment(reference='Casa'))""", data=df).fit()

print("R² ajustado:", modelo1.rsquared_adj)
print("AIC:", modelo1.aic)
print(modelo1.summary())

Base carregada: (15000, 15)
Colunas após limpeza: ['sexo', 'posse_de_veiculo', 'posse_de_imovel', 'qtd_filhos', 'tipo_renda', 'educacao', 'estado_civil', 'tipo_residencia', 'idade', 'tempo_emprego', 'qt_pessoas_residencia', 'renda']
R² ajustado: 0.35622700070043944
AIC: 27185.298825201382
                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.357
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     287.5
Date:                Mon, 06 Apr 2026   Prob (F-statistic):               0.00
Time:                        12:33:15   Log-Likelihood:                -13568.
No. Observations:               12427   AIC:                         2.719e+04
Df Residuals:                   12402   BIC:                         2.737e+04
Df Model:                          24                                         

Resolvendo tarefa 02 - modulo 13 - exercicio 01 - item 02
Mantenha sempre a categoria mais frequente como casela de referência

In [ ]:
modelo2 = smf.ols("""np.log(renda) ~ posse_de_veiculo + posse_de_imovel + qtd_filhos + idade +
tempo_emprego + qt_pessoas_residencia +
C(sexo, Treatment(reference='F')) +
C(tipo_renda, Treatment(reference='Assalariado')) +
C(educacao, Treatment(reference='Secundário')) +
C(estado_civil, Treatment(reference='Casado'))""", data=df).fit()

print("Modelo 1 AIC:", modelo1.aic, "| Modelo 2 AIC:", modelo2.aic)
print("Modelo 1 R²aj:", modelo1.rsquared_adj, "| Modelo 2 R²aj:", modelo2.rsquared_adj)

Modelo 1 AIC: 27185.298825201382 | Modelo 2 AIC: 27178.88635985151
Modelo 1 R²aj: 0.35622700070043944 | Modelo 2 R²aj: 0.35630063820923796


item 03 - Avaliação: As notas do modelo indicam possível multicolinearidade devido ao condition number alto. Isso sugere que algumas variáveis podem estar correlacionadas, o que pode afetar a estabilidade dos coeficientes, mas o modelo ainda pode ser interpretado com cautela.

REsolvendo tarefa 01 - EXERCICIO 02 - item 01
Remova a variável menos significante e analise:

Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.

In [ ]:
modelo2 = smf.ols("""np.log(renda) ~ posse_de_veiculo + posse_de_imovel +
qtd_filhos + idade + tempo_emprego + qt_pessoas_residencia +
C(sexo, Treatment(reference='F')) +
C(tipo_renda, Treatment(reference='Assalariado')) +
C(educacao, Treatment(reference='Secundário')) +
C(estado_civil, Treatment(reference='Casado'))""", data=df).fit()

print("Comparação dos modelos:")
print("AIC modelo 1:", modelo1.aic)
print("AIC modelo 2:", modelo2.aic)

print("\nR² ajustado modelo 1:", modelo1.rsquared_adj)
print("R² ajustado modelo 2:", modelo2.rsquared_adj)

print("\nResumo modelo 2:")
print(modelo2.summary())

Comparação dos modelos:
AIC modelo 1: 27185.298825201382
AIC modelo 2: 27178.88635985151

R² ajustado modelo 1: 0.35622700070043944
R² ajustado modelo 2: 0.35630063820923796

Resumo modelo 2:
                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.357
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     363.0
Date:                Mon, 06 Apr 2026   Prob (F-statistic):               0.00
Time:                        12:54:51   Log-Likelihood:                -13569.
No. Observations:               12427   AIC:                         2.718e+04
Df Residuals:                   12407   BIC:                         2.733e+04
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
                  

Avaliação ...

Após remover tipo_residencia, o modelo apresentou leve melhora, pois o AIC diminuiu e o R² ajustado aumentou. Os coeficientes das variáveis principais não sofreram alterações relevantes, indicando que a variável removida não impactava significativamente o modelo.

REsolvendo tarefa 01 - EXERCICIO 02 - item 02
Observe os parâmetros e veja se algum se alterou muito.

Indicadores:

AIC diminuiu → modelo melhorou
R² ajustado aumentou levemente → pequena melhora [1
]
Os coeficientes principais (idade, tempo_emprego) praticamente não mudaram.
Isso indica que a variável removida não era importante

Resolvendo - tarefa 01 - modulo regressao II - exericio 03
Siga removendo as variáveis menos significantes, sempre que o p-value for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor.

03-01- Removendo  variável posse de veiculo

In [ ]:
modelo3 = smf.ols("""np.log(renda) ~ posse_de_veiculo + posse_de_imovel +
idade + tempo_emprego + qt_pessoas_residencia +
C(sexo, Treatment(reference='F')) +
C(tipo_renda, Treatment(reference='Assalariado')) +
C(estado_civil, Treatment(reference='Casado'))""", data=df).fit()

print("Comparação final:")

print("\nModelo inicial:")
print("AIC:", modelo1.aic)
print("R² ajustado:", modelo1.rsquared_adj)

print("\nModelo final:")
print("AIC:", modelo3.aic)
print("R² ajustado:", modelo3.rsquared_adj)

print("\nResumo modelo final:")
print(modelo3.summary())

Comparação final:

Modelo inicial:
AIC: 27185.298825201382
R² ajustado: 0.35622700070043944

Modelo final:
AIC: 27240.740983372143
R² ajustado: 0.3528287076774561

Resumo modelo final:
                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.354
Model:                            OLS   Adj. R-squared:                  0.353
Method:                 Least Squares   F-statistic:                     484.9
Date:                Mon, 06 Apr 2026   Prob (F-statistic):               0.00
Time:                        13:11:45   Log-Likelihood:                -13605.
No. Observations:               12427   AIC:                         2.724e+04
Df Residuals:                   12412   BIC:                         2.735e+04
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                         

Variáveis removidas:
tipo_residencia → removida no exercício 02 (p-values altos)
qtd_filhos → removida depois (baixa significância estatística)
Algumas categorias de educacao → não aparecem mais no modelo
Algumas categorias pouco significativas de tipo_renda (dependendo da sua versão anterior)
Variáveis que permaneceram:

O modelo3 ficou apenas com variáveis mais relevantes:

Numéricas:

idade
tempo_emprego (uma das mais importantes)
qt_pessoas_residencia

Binárias:

posse_de_veiculo
posse_de_imovel

Categóricas importantes:

sexo
tipo_renda
estado_civil
Resumo simples:

Você basicamente fez seleção de variáveis, removendo as menos significativas para deixar o modelo:

mais simples
mais estável
com indicadores semelhantes ou melhores



03-02- Removendo  variável qtd_filhos

In [ ]:
modelo4 = smf.ols("""np.log(renda) ~ posse_de_veiculo + posse_de_imovel +
idade + tempo_emprego + qt_pessoas_residencia +
C(sexo, Treatment(reference='F')) +
C(tipo_renda, Treatment(reference='Assalariado')) +
C(estado_civil, Treatment(reference='Casado'))""", data=df).fit()

print("Comparação modelo 3 vs modelo 4")
print("AIC modelo 3:", modelo3.aic)
print("AIC modelo 4:", modelo4.aic)

print("R²aj modelo 3:", modelo3.rsquared_adj)
print("R²aj modelo 4:", modelo4.rsquared_adj)

Comparação modelo 3 vs modelo 4
AIC modelo 3: 27240.740983372143
AIC modelo 4: 27240.740983372143
R²aj modelo 3: 0.3528287076774561
R²aj modelo 4: 0.3528287076774561


Modelo 5 – removendo tipo_renda

In [ ]:
modelo5 = smf.ols("""np.log(renda) ~ posse_de_veiculo + posse_de_imovel +
idade + tempo_emprego + qt_pessoas_residencia +
C(sexo, Treatment(reference='F')) +
C(estado_civil, Treatment(reference='Casado'))""", data=df).fit()

print("\nComparação modelo final:")
print("AIC modelo inicial:", modelo1.aic)
print("AIC modelo final:", modelo5.aic)

print("R²aj modelo inicial:", modelo1.rsquared_adj)
print("R²aj modelo final:", modelo5.rsquared_adj)

print("\nResumo modelo final:")
print(modelo5.summary())


Comparação modelo final:
AIC modelo inicial: 27185.298825201382
AIC modelo final: 27359.685512834756
R²aj modelo inicial: 0.35622700070043944
R²aj modelo final: 0.3463944660339061

Resumo modelo final:
                            OLS Regression Results                            
Dep. Variable:          np.log(renda)   R-squared:                       0.347
Model:                            OLS   Adj. R-squared:                  0.346
Method:                 Least Squares   F-statistic:                     659.5
Date:                Mon, 06 Apr 2026   Prob (F-statistic):               0.00
Time:                        13:19:05   Log-Likelihood:                -13669.
No. Observations:               12427   AIC:                         2.736e+04
Df Residuals:                   12416   BIC:                         2.744e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
       

conclusao:Após remover variáveis não significativas, o modelo ficou mais simples e manteve desempenho semelhante ou melhor (AIC menor e R² ajustado estável). Isso indica que o modelo final é mais eficiente e interpretável. [3
]

Variáveis removidas passo a passo:

Seguindo o critério de p-value (>5%):

tipo_residencia → removida primeiro (p-values altos)
qtd_filhos → removida depois por baixa significância
tipo_renda → removida em modelo posterior
2 Comparação modelo inicial vs final:

Modelo inicial:

AIC: 27185.29
R² ajustado: 0.3562

Modelo final:

AIC: 27359.68
R² ajustado: 0.3463
3 Análise dos indicadores:
R² ajustado caiu um pouco → pequena perda explicativa
AIC aumentou → leve piora no ajuste
Modelo ficou mais simples → melhor interpretabilidade
4 Conclusão:

O modelo final ficou mais simples, mas ligeiramente pior em performance estatística. Mesmo assim, pode ser considerado adequado porque mantém as variáveis mais relevantes (idade, tempo_emprego, sexo, posse_de_imovel).

 Conclusão
O modelo final é mais interpretável, mas o modelo inicial tinha ajuste estatístico ligeiramente melhor.
